### PR TITLE
Make tmp dir configurable and dynamic at local-up-cluster.sh

### LINKED ISSUE #115016: Permission errors at local-up-cluster.sh on /tmp directory
### What happened?

I tried to execute the script to start a cluster and it failed:

```
./hack/local-up-cluster.sh: line 515: /tmp/kube_egress_selector_configuration.yaml: Permission denied
./hack/local-up-cluster.sh: line 533: /tmp/kube-audit-policy-file: Permission denied
```

Update: After some investigation, I have found that I executed the script without `sudo` first (my bad) and it failed with sudo next time.

### What did you expect to happen?

The script starts a cluster without any temp folder-related permission issues.

### How can we reproduce it (as minimally and precisely as possible)?

Run `local-up-cluster.sh` without `sudo` first "accidentally" and then re-execute the command with `sudo`.

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"27+", GitVersion:"v1.27.0-alpha.0.908+a887a3b4fdebe0-dirty", GitCommit:"a887a3b4fdebe0390c6019b5fc5eb911e8042cec", GitTreeState:"dirty", BuildDate:"1980-01-01T00:00:00Z", GoVersion:"go1.19.5", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"27+", GitVersion:"v1.27.0-alpha.0.908+a887a3b4fdebe0-dirty", GitCommit:"a887a3b4fdebe0390c6019b5fc5eb911e8042cec", GitTreeState:"dirty", BuildDate:"1980-01-01T00:00:00Z", GoVersion:"go1.19.5", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
BUG_REPORT_URL="https://github.com/NixOS/nixpkgs/issues"
BUILD_ID="22.11.1482.8c54d842d95"
DOCUMENTATION_URL="https://nixos.org/learn.html"
HOME_URL="https://nixos.org/"
ID=nixos
LOGO="nix-snowflake"
NAME=NixOS
PRETTY_NAME="NixOS 22.11 (Raccoon)"
SUPPORT_END="2023-06-30"
SUPPORT_URL="https://nixos.org/community.html"
VERSION="22.11 (Raccoon)"
VERSION_CODENAME=raccoon
VERSION_ID="22.11"
$ uname -a
Linux nixos 5.15.86 #1-NixOS SMP Sat Dec 31 12:14:48 UTC 2022 x86_64 GNU/Linux
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by mhmxs**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Permissions of the generated config files are too strictly created by `local-up-cluster.sh` and this should lead to permission errors.

```
./hack/local-up-cluster.sh: line 515: /tmp/kube_egress_selector_configuration.yaml: Permission denied
./hack/local-up-cluster.sh: line 533: /tmp/kube-audit-policy-file: Permission denied
```

This change generates a separate `tmp` directory for every cluster created by the script.

#### Which issue(s) this PR fixes:

Fixes #115016

#### Special notes for your reviewer:

I have reused the command from: https://github.com/kubernetes/kubernetes/blob/0ff0d0b94e49396724b790aa073d62e6152e1cad/hack/lib/etcd.sh#L83

#### Does this PR introduce a user-facing change?

No

### CODE REVIEW

**Comment by mhmxs** [comment]
Does it make sense to use the more common `TMPDIR` environment variable instead? For example, Go uses `TMPDIR` during compilation.

**Comment by mhmxs** [comment]
This change isn't relevant, because other scripts are also using the hard coded `/tmp` folder.

**Comment by mhmxs** [comment]
Based on the issue: https://github.com/kubernetes/kubernetes/pull/115106 I think this is still relevant, so makes sense to reopen.
FYI: @bart0sh

File: hack/local-up-cluster.sh [hunk 1]
```diff
@@ -217,6 +217,7 @@ LOG_LEVEL=${LOG_LEVEL:-3}
 # Use to increase verbosity on particular files, e.g. LOG_SPEC=token_controller*=5,other_controller*=4
 LOG_SPEC=${LOG_SPEC:-""}
 LOG_DIR=${LOG_DIR:-"/tmp"}
+TMP_DIR=${TMP_DIR:-$(mktemp -d 2>/dev/null || mktemp -d -t test-cluster.XXXXXX)}
```
Comment (bart0sh) [inline hack/local-up-cluster.sh:220]: This approach still needs to be understood. if we don't find proper explanation of why it's used, I'd use `TMP_DIR=${TMP_DIR:-$(mktemp -d -t test-cluster.XXXXXXXXXX)}`

Comment (bart0sh) [reply hack/local-up-cluster.sh:220]: @mhmxs let's use [this approach](https://github.com/kubernetes/kubernetes/pull/115106#discussion_r1072441528) suggested by @thockin

Comment (mhmxs) [reply hack/local-up-cluster.sh]: I did @bart0sh, thanks for the review

**Review by thockin — COMMENTED** [review]
Thanks!

/lgtm
/approve

**Pull request merged**